kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java

```diff
@@ -2258,34 +2258,35 @@ protected boolean isTransactionTopic(String topic) {
         CompletableFuture<PartitionMetadata> returnFuture = new CompletableFuture<>();
 
         topicManager.getTopicBroker(topic.toString())
-            .thenCompose(address -> getProtocolDataToAdvertise(address, topic))
-            .whenComplete((stringOptional, throwable) -> {
-                if (!stringOptional.isPresent() || throwable != null) {
-                    log.error("Not get advertise data for Kafka topic:{}. throwable: [{}]",
-                        topic, throwable.getMessage());
-                    KafkaTopicManager.removeTopicManagerCache(topic.toString());
-                    returnFuture.complete(null);
-                    return;
-                }
+                .thenCompose(address -> getProtocolDataToAdvertise(address, topic))
+                .whenComplete((stringOptional, throwable) -> {
```

Comment:
Combining `thenCompose` and `whenComplete` is not a good way to handle future. The null completed value is confused that it looks like a `Optional<T>` future might complete with null value.

It's better to combine `thenXXX` with `exceptionally`, with could handle all possible exceptions in previous chains.

```java
future.thenApply(/* block 1 */).thenApply(/* block 2 */).exceptionally(e -> {
    // e could be from block 1 or block 2 or even future itself
    // if e was thrown from block 1 or block 2, e's type is CompletionException, e.getCause() is the actual exception
    return null;
});
```

`whenComplete` is mostly used when you just need a callback for a future.

Reply:
Good suggestion. I already used `thenApply` to replase `thenCompose` now. PTAL.